onnxruntime/core/framework/fallback_cpu_capability.cc

```diff
@@ -39,6 +43,132 @@ static bool IsSmallInitializer(const onnxruntime::GraphViewer& graph, const Node
 }
 }  // namespace
 
+std::unordered_set<NodeIndex> GetShapeRelatedNodes(const onnxruntime::GraphViewer& viewer) {
+  // Conceptually, this function traverse from shape-consuming nodes
+  // to fallback all its upstream nodes to CPU. Consider a graph
+  //
+  //
+  // The traversal should stop when
+  //  1. hitting Shape, Size nodes, graph inputs, or graph initializers.
+  //  2. hitting nodes with some large inputs or outputs.
```

Comment:
The comment says 'traversal should stop when hitting nodes with some large inputs or outputs' - is there logic below to capture this intuition ?

Reply:
Thanks for capturing this. The code checked shapes before but there are a bunch of nodes without shapes and the check is removed. I will add a TODO, which requires shape inference before calling this function.